src/main/java/hudson/remoting/RemoteClassLoader.java

```diff
@@ -130,10 +133,48 @@ private RemoteClassLoader(@CheckForNull ClassLoader parent, @Nonnull IClassLoade
         final Channel channel = RemoteInvocationHandler.unwrap(proxy);
         this.channel = channel == null ? null : channel.ref();
         this.underlyingProxy = proxy;
+        final IClassLoader targetProxy;
         if (channel == null || !channel.remoteCapability.supportsPrefetch() || channel.getJarCache()==null) {
-            proxy = new DumbClassLoaderBridge(proxy);
+            targetProxy = new DumbClassLoaderBridge(proxy);
+        } else {
+            targetProxy = proxy;
         }
-        this.proxy = proxy;
+        // JENKINS-61103: wrap the target (IClassLoader) proxy in a dynamic proxy which will handle retries in case of
+        // interrupt exception from the remoting system during class loading/initialization.
+        // It should take care of cases like JENKINS-36991, JENKINS-51854 and others.
+        this.proxy = (IClassLoader) Proxy.newProxyInstance(RemoteClassLoader.class.getClassLoader(),
```

Comment:
My biggest concern with this approach is the deep nesting and how it impacts readability. There are a number of different options, but we should try something to improve readability.

My next concern is that I have been bitten in the past by performance degradation from reflection. Most of the time it's not enough to cause concern, but I'm a little worried about what it might do in this area.

I like how this handles it at a central, localized place, but I think it would be worth exploring if it can be done in a more readable fashion without reflection performance concerns using something different from a dynamic proxy here.

Reply:
85a19dee is an attempt at improving readability. The `InvocationHandler` is not lambda-defined anymore, it has comments, and the logic to decide whether there could be retries given a method and a caugh interruption is not buried in the middle of the `if` condition anymore (that's `RemoteClassLoader.RemoteInterruptedExceptionCatcher.shouldRetry(Method,Throwable)`).

This is still using a dynamic proxy though. I've not found time to work more on this and experiment alternative solutions, or add jmh benchmarks. So that's all I have for now, I thought it's better to share despite it not being an exhaustive answer to your concerns, because time flies. 

I'm not super worried about the impact of basing implementation on a dynamic proxy actually; it is "only" adding one more layer in calls to an other dynamic proxy (`RemoteInvocationHandler`), so at worst it's doubling an already existing penalty :)

But yes, would be good to explore perfs and alternatives, I just can't tell when I will be able to do it.